File: aufgabe2.py
import numpy as np
import itertools 
from matplotlib import pyplot as plt 
import matplotlib.cm as cm
#from mpl_toolkits.mplot3d import Axes3D
#fig = plt.figure()
#ax = fig.add_subplot(111, projection='3d')



















































































# nb = 64
# xA = np.genfromtxt("Aufgabe2bxAN.txt")
# yA = np.genfromtxt("Aufgabe2byAN.txt")

# zg = np.genfromtxt("Aufgabe2a.txt")

# fig, ax1 = plt.subplots()
# ax1.hist(zg,nb,lw=1,linestyle='solid',edgecolor='black',color='blue')
# ax1.tick_params('y', colors='b')
# ax1.set_xlabel('z')
# ax1.set_ylabel('N', color='b')
# plt.title("Gaußverteilte Zahlen mit Box-Muller")

# ax2 = ax1.twinx()
# ax2.set_position([0.125,0.076,0.775,0.803])
# ax2.set_frame_on(False)
# ax2.plot(xA,yA,lw=1,color='r')
# ax2.tick_params('y', colors='r')
# ax2.set_xlabel('x')
# ax2.set_ylabel('f(x)', color='r')
# plt.savefig("Aufgabe2a.pdf")
# plt.close()

# zwg = np.genfromtxt("Aufgabe2b.txt")



# fig, ax1 = plt.subplots()
# ax1.hist(zwg,nb,lw=1,linestyle='solid',edgecolor='black',color='blue')
# ax1.tick_params('y', colors='b')
# ax1.set_xlabel('z')
# ax1.set_ylabel('N', color='b')
# plt.title("Gaußverteilte Zahlen mit zentralen Grenzwertsatz")

# ax2 = ax1.twinx()
# ax2.set_position([0.125,0.076,0.775,0.803])
# ax2.set_frame_on(False)
# ax2.plot(xA,yA,lw=1,color='r')
# ax2.tick_params('y', colors='r')
# ax2.set_xlabel('x')
# ax2.set_ylabel('f(x)', color='r')
# plt.savefig("Aufgabe2b.pdf")
# plt.close()

# ###################################################
# ###################################################
# ###################################################

# zNEU = np.genfromtxt("Aufgabe2c.txt")

# fig, ax1 = plt.subplots()
# ax1.hist(zNEU,nb,lw=1,linestyle='solid',edgecolor='black',color='blue')
# ax1.tick_params('y', colors='b')
# ax1.set_xlabel('z')
# ax1.set_ylabel('N', color='b')
# #plt.title("Gaußverteilte Zahlen mit Neumann")

# ax2 = ax1.twinx()
# ax2.set_position([0.125,0.076,0.775,0.803])
# ax2.set_frame_on(False)
# ax2.plot(xA,yA,lw=1,color='r')
# ax2.tick_params('y', colors='r')
# ax2.set_xlabel('x')
# ax2.set_ylabel('f(x)', color='r')
# plt.savefig("Aufgabe2c.pdf")
# plt.close()



# zVNEU = np.genfromtxt("Aufgabe2c2.txt")
# a = np.array([0.01,0.1,0.5,1.0,1.5,10])

# fig, axes = plt.subplots(2,3,squeeze=False)#,sharey=True)
# fig.set_size_inches(14,7)

# axes[0,0].hist(zVNEU[0],nb,lw=1,linestyle='solid',edgecolor='black',color='blue',normed=True)
# axes[0,0].tick_params('y', colors='b')
# axes[0,0].set_xlabel('z')
# axes[0,0].set_ylabel('N bzw. f(x)', color='b')
# axes[0,0].plot(xA,yA,lw=1,color='r')
# axes[0,0].set_title(r"Gaußverteilte Zahlen mit Neumann und $\alpha = 0.01$")


# plt.title("Gaußverteilte Zahlen mit Neumann")
# axes[0,1].hist(zVNEU[1],nb,lw=1,linestyle='solid',edgecolor='black',color='blue',normed=True)
# axes[0,1].tick_params('y', colors='b')
# axes[0,1].set_xlabel('z')
# axes[0,1].set_ylabel('N bzw. f(x)', color='b')
# axes[0,1].plot(xA,yA,lw=1,color='r')
# axes[0,1].set_title(r"Gaußverteilte Zahlen mit Neumann und $\alpha = 0.1$")



# axes[0,2].hist(zVNEU[2],nb,lw=1,linestyle='solid',edgecolor='black',color='blue',normed=True)
# axes[0,2].tick_params('y', colors='b')
# axes[0,2].set_xlabel('z')
# axes[0,2].set_ylabel('N bzw. f(x)', color='b')
# axes[0,2].plot(xA,yA,lw=1,color='r')
# axes[0,2].set_title(r"Gaußverteilte Zahlen mit Neumann und $\alpha = 1.0$")


# axes[1,0].hist(zVNEU[3],nb,lw=1,linestyle='solid',edgecolor='black',color='blue',normed=True)
# axes[1,0].tick_params('y', colors='b')
# axes[1,0].set_xlabel('z')
# axes[1,0].set_ylabel('N bzw. f(x)', color='b')
# axes[1,0].plot(xA,yA,lw=1,color='r')
# axes[1,0].set_title(r"Gaußverteilte Zahlen mit Neumann und $\alpha = 1.5$")



# axes[1,1].hist(zVNEU[4],nb,lw=1,linestyle='solid',edgecolor='black',color='blue',normed=True)
# axes[1,1].tick_params('y', colors='b')
# axes[1,1].set_xlabel('z')
# axes[1,1].set_ylabel('N bzw. f(x)', color='b')
# axes[1,1].plot(xA,yA,lw=1,color='r')
# axes[1,1].set_title(r"Gaußverteilte Zahlen mit Neumann und $\alpha = 10$")



# axes[1,2].hist(zVNEU[5],nb,lw=1,linestyle='solid',edgecolor='black',color='blue',normed=True)
# axes[1,2].tick_params('y', colors='b')
# axes[1,2].set_xlabel('z')
# axes[1,2].set_ylabel('N bzw. f(x)', color='b')
# axes[1,2].plot(xA,yA,lw=1,color='r')
# axes[1,2].set_title(r"Gaußverteilte Zahlen mit Neumann und $\alpha = 100$")

# plt.tight_layout()

# plt.savefig("Aufgabe2c2.pdf")
# plt.close()


# zinv = np.genfromtxt("Aufgabe2d.txt")
# xANd = np.genfromtxt("Aufgabe2dxAN.txt")
# yANd = np.genfromtxt("Aufgabe2dyAN.txt")
# zinv = zinv[np.absolute(zinv) < 100]


# fig, ax1 = plt.subplots()
# ax1.hist(zinv,60,lw=1,linestyle='solid',edgecolor='black',color='blue')
# ax1.tick_params('y', colors='b')
# ax1.set_xlabel('z')
# ax1.set_ylabel('N', color='b')
# plt.title("Zufallszahlen mit Inversionsmethode")


# ax2 = ax1.twinx()
# ax2.set_position([0.125,0.076,0.775,0.803])
# ax2.set_frame_on(False)
# ax2.plot(xANd,yANd,lw=1,color='r')
# ax2.tick_params('y', colors='r')
# ax2.set_xlabel('x')
# ax2.set_ylabel('f(x)', color='r')
# plt.savefig("Aufgabe2d.pdf")
# plt.close()


# zinvSMD = np.genfromtxt("Aufgabe2dSMD.txt")
# zinvSMD = zinvSMD[np.absolute(zinvSMD) < 100]

# plt.hist(zinvSMD,1500,lw=0.5,linestyle='solid',edgecolor='black',color='blue',normed=True,stacked=True)
# plt.plot(xANd,yANd,color="red")
# plt.xlabel("z")
# plt.ylabel('N bzw. f(x)')
# plt.xlim(-4.5,4.5)
# plt.title("Zufallszahlen mit Inversionsmethode aus SMD")
# plt.savefig("Aufgabe2dSMD.pdf")

File: aufgabe2.cpp
#include "../../eigen/Eigen/Dense"
#include <iostream>
#include <fstream>
#include <math.h> 
#include <cmath>
#include <functional>
#include <random>
#include "functions.hpp"

using Eigen::MatrixXd;
using Eigen::MatrixXcd;
using Eigen::VectorXd;
using Eigen::VectorXcd;
using std::cout;
using std::endl;
using std::ofstream;
using std::complex;
using std::function;



int main()
{
	
	return 0;
}
























// //Box-Muller-Algorithmus
// VectorXd boxm(int n);
// //Zentraler Grenzwertsatz Funktion flexible sigma Eingabe
// VectorXd zgw(int n, double sigma);
// //Normaler Gauß halt
// double gaus(double x, double mu, double sigma);
// //Neumann Algorithmus um eine Gauß verteilung zu erhalten, dabei wird die Fläche angepasst siehe unten
// VectorXd neumannGAUS(int n, double mu, double sigma, double alpha, function < double(double,double,double)  > f);
// //Transformationsverfahren aus Kierfeld Sktipt
// VectorXd inversionp(int n);
// //Transformationsverfahren aus SMD
// VectorXd inversionpSMD(int n);
// //Die gesuchte Verteilung für d)
// double p(double x);

// int main(){
// 	int n = 1e5;	//Wir wollen immer 10^5 Zahlen

// 	VectorXd zg = boxm(n);		//Box-Muller für a)
// 	speichere("Aufgabe2a",zg);

// 	VectorXd Zzgw = zgw(n,1);	//Zentraler Grenzwertsatz für b)
// 	speichere("Aufgabe2b",Zzgw);

// 	VectorXd xAN(n),yAN(n);		//Anlytischer Gauß für Plot
// 	for (int i = 0; i < n; ++i)
// 	{
// 		xAN[i] = ((1.*i)/(1.*n))*10-5;	//Benutze den Gauß zwischen -5 und 5 (sollte hier reichen)
// 		yAN[i] = gaus(xAN[i],0,1);
// 	}
// 	speichere("Aufgabe2bxAN",xAN);
// 	speichere("Aufgabe2byAN",yAN);

// 	//Das hier ist nur zur Kontrolle gewesen mit einem alpha = 1.5 (wie im skript)
// 	VectorXd zNEU(n);
// 	double alpha = 1.5;
// 	zNEU = neumannGAUS(n,0,1,alpha,gaus);	
// 	speichere("Aufgabe2c",zNEU);

// 	//Berechne die Gaußverteilung mit Neumann für verschiedene alphas 
// 	MatrixXd zaNEU(6,n);
// 	VectorXd alphaV(6);

// 	alphaV << 0.01, 0.1, 1.0, 1.5, 10, 100;

// 	for (int i = 0; i < 6; ++i)
// 	{
// 		zaNEU.row(i) = neumannGAUS(n,0,1,alphaV[i],gaus);
// 	}
// 	speichere("Aufgabe2c2",zaNEU);



// 	VectorXd zinv(n);		//Inversionsmethode aus Skript
// 	zinv = inversionp(n);
// 	speichere("Aufgabe2d",zinv);

// 	//Analytische Verteilung aus d)
// 	VectorXd xANd(n),yANd(n);
// 	for (int i = 0; i < n; ++i)
// 	{
// 		xANd[i] = ((1.*i)/(1.*n))*20-10;	//Zwischen -5 und 5 Werte für p(x)
// 		yANd[i] = p(xANd[i]);
// 	}
// 	speichere("Aufgabe2dxAN",xANd);
// 	speichere("Aufgabe2dyAN",yANd);

// 	VectorXd zinvSMD(n);			//Andere Inversionsmethode
// 	zinvSMD = inversionpSMD(n);
// 	speichere("Aufgabe2dSMD",zinvSMD);

// 	return 0;
// }

// //Box-Muller Algorithmus
// VectorXd boxm(int n){
// 	VectorXd x1(n),x2(n),z1(n),z2(n);

// 	std::random_device rd;  
//     std::mt19937 gen(rd()); 
//     std::uniform_real_distribution<> dis(0, 1);

//     //Ziehe und speichere Zunächst 2 Zufallszahlen (jeweils n Stück)
//     for (int i = 0; i < n; ++i)
//     {
//     	x1[i] = dis(gen);
//     	x2[i] = dis(gen);
//     }
//     //Nutze Box-Muller zum Erzeugen der Gauß Verteilungen
//     for (int i = 0; i < n; ++i)
//     {
//     	z1[i] = sqrt(-2*log(x1[i]))*cos(2*M_PI*x2[i]);
//     	z2[i] = sqrt(-2*log(x1[i]))*sin(2*M_PI*x2[i]);
//     }
//     return z1;	//Übergebe einen der beiden (evtl wäre Mischen effizienter aber die Zahlen könnten korreliert sein...)
// }

// //Gauß mit Zentralen Grenzwertsatz
// VectorXd zgw(int n, double sigma){

// 	//Berechne zunächst wie viele Zahlen aufsummiert werden müssen
// 	//Das ist von Sigma abhängig und wird aufgerundet abgeschnitten für Int
// 	int N = int((pow(sigma,2)*12)+0.5); 

// 	//z sind später die Zufallszahlen und y hilfsvariable zum summieren
// 	VectorXd z(n);
// 	double y;

// 	std::random_device rd;  
//     std::mt19937 gen(rd()); 
//     std::uniform_real_distribution<> dis(0, 1);

//     //Gehe durch n (Anzahl an gewollten Zufallszahlen)
//     for (int i = 0; i < n; ++i)
//     {
//     	y = 0;							//Setze zunächst y = 0
//     	for (int j = 0; j < N; ++j)
//     	{
//     		y += dis(gen);				//Summiere N Zufallszahlen [0,1] auf
//     	}
//     	y -= N/2.;						//Ziehe N/2 ab (s.h. Skript)
//     	z[i] = y;						//Abspeichern
//     }
//     return z;							//Zurückgeben
// }

// //Üblicher Gauß mit x,mu,sigma als Input
// double gaus(double x, double mu, double sigma){
// 	return 1./(sqrt(2*M_PI)*sigma) * exp((-1.*pow((x-mu),2))/(2*pow(sigma,2)));
// }


// //Gauß mit Neumann Ziehen, hier kann theoretisch auch eine andere Funktion der form double(double,double,double)
// //erzeugt werden, jedoch ist zu beachten dass die Fläche in der Zahlen gezogen Werden sehr angepasst an gauß ist 
// VectorXd neumannGAUS(int n, double mu, double sigma, double alpha, function < double(double,double,double)  > f){
	
// 	std::random_device rd;
//     std::mt19937 gen(rd());
 	
//  	//Ziehe hier exponential verteilte Zahlen hier kommen zahlen der form 1 * exp(-1*x) heraus
//     std::exponential_distribution<> d(1);
// 	std::uniform_real_distribution<> dis(0, 1);
//     //z zum Abspeichern und zexp und zy als Hilfsvariablen
// 	VectorXd z(n);
// 	double zexp,zy;

// 	//i_n Zählt wie viele Zahlen wir durch Neumann haben damit wir auch 10^5 Stück bekommen
// 	int i_n = 0;
// 	do
// 	{
// 		//Ziehe zunächst einen Exponentialverteilung für x Wert
// 		//dabei wird eine weitere Zahl gezogen um auch den negativen Bereich abzudecken
// 		zexp =  (dis(gen) > 0.5)?(-1)*d(gen):d(gen); 
// 		//Ziehe in y Richtung gleichverteilte Zahlen gewichtet mit der Funktion die gefordert war um den Definitionsberech abzudeken
// 		zy = dis(gen)*alpha*exp(-fabs(zexp));
// 		//wenn die y Zahl in dem Gauß liegt akzeptieren sonst gehts weiter...
// 		if(zy <= f(zexp,mu,sigma))
// 		{
// 			z[i_n] = zexp;
// 			if(i_n < n){i_n++;}
// 		}
// 	}while(i_n < n);

// 	return z;
// }

// //Die gesuchte Verteilung
// double p(double x){
// 	return (1/M_PI)*(1/(1+pow(x,2)));
// }

// //Die inversen der gesuchten Verteilung
// double pi(double y){
// 	return sqrt((1/(M_PI*y))-1);
// }

// //Die Ableitung der inversen Verteilung
// double dpi(double y){
// 	return -1/(2*M_PI*pow(y,2)*sqrt((1/(M_PI*y))-1));
// }

// //Inversionsmethode aus dem Kierfeld Skript
// VectorXd inversionp(int n){
// 	std::random_device rd;  
//     std::mt19937 gen(rd()); 
//     std::uniform_real_distribution<> dis(0, 1);

//     VectorXd z(n),zN(n);
//     double y;

//     for (int i = 0; i < n; ++i)
//     {
//     	//Ziehe zufällig gleichverteilt im Intervall 0 bis 1/pi
//     	y = dis(gen)*((1/M_PI));
//     	zN[i] = (dis(gen) > 0.5)?((-1)*fabs(dpi(y))):(fabs(dpi(y)));
//     }
//     return zN;
// }

// //Inversionsmethode aus SMD
// VectorXd inversionpSMD(int n){
// 	std::random_device rd;  
//     std::mt19937 gen(rd()); 
//     std::uniform_real_distribution<> dis(0, 1);

//     VectorXd z(n),zN(n);
//     double y;

//     for (int i = 0; i < n; ++i)
//     {
//     	//Ziehe Zufallszahlen nach (z-1/2) also zwischen -1/2 und 1/2 und einsetzten -> s.h. getechte Formel
//     	y = dis(gen)-0.5;
//     	zN[i] = tan(M_PI*y);
//     }
//     return zN;
// }